stablehlo/dialect/TypeInference.cpp

```diff
@@ -802,6 +802,63 @@ LogicalResult inferDotGeneralOp(
   return success();
 }
 
+LogicalResult inferDynamicUpdateSliceOp(
+    Optional<Location> location, Value operand, Value update,
+    ValueRange startIndices,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  auto operandType = operand.getType().cast<ShapedType>();
+  auto updateType = update.getType().cast<ShapedType>();
+
+  // (C2)
+  if (updateType.getElementType() != operandType.getElementType())
+    return emitOptionalError(
+        location, "update element type does not match operand element type: ",
+        updateType.getElementType(), " vs ", operandType.getElementType(), ".");
```

Comment:
Please add a test for C2, and we're good to go.

Reply:
I deliberately did not add a test for this since the ODS already enforces it:
```
def StableHLO_DynamicUpdateSliceOp: StableHLO_Op<"dynamic_update_slice",
      [Pure, AllElementTypesMatch<["operand", "update", "result"]>,
       AllShapesMatch<["operand", "result"]>, InferTensorType]> {
```

Reply:
I did reorder the tests to follow the ordering of the constraints, which it wasn't before.

Reply:
Oh! I missed that part. Since C2 is covered by ODS, we don't need a check in the code.